### PR TITLE
Links

### PR DESCRIPTION
**Pull request by nschan**
This is a module for the scaffolding tool [`LINKS`](https://github.com/bcgsc/LINKS), part of the [genomeassembler local module migration](https://github.com/nf-core/genomeassembler/issues/110).
I am not sure if the singularity container actually works, locally it produced errors. This tool has another limitation, it cannot do more than 3 (or 4, there is conflicting information in the README and issue comments) threads, without crashing. 

log and graph are excluded from the test, the log contains timestamps, the graph seems to be somewhat stochastic.

### CODE REVIEW

**Comment by nschan** [comment]
Looks like the singularity container works fine in a runner.

File: modules/nf-core/links/tests/main.nf.test [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                { assert process.success },
+                { assert snapshot(
+                    process.out.pairing_issues,
+                    process.out.scaffolds_csv,
+                    process.out.scaffolds_fasta,
+                    process.out.bloom,
+                    process.out.assembly_correspondence,
+                    process.out.tigpair_checkpoint,
+                    process.out.versions
+                    ).match()
```
Comment (maxulysse) [inline modules/nf-core/links/tests/main.nf.test:44]: could you assert the filenames of the output you're not capturing the output?

File: modules/nf-core/links/tests/main.nf.test [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+        when {
+            params {
+                module_args = "-d 1000,2000,3000,4000,5000,6000,7000,8000,9000,10000,12000,14000,16000,18000,20000"
+            }
+            process {
+                """
+                input[0] = [
+                    [ id:'test', single_end:false ], // meta map
+                    file(params.modules_testdata_base_path + 'delete_me/LINKS/contigs.fasta', checkIfExists: true),
```
[the remark this answers is not in the document]
Comment (nschan) [reply modules/nf-core/links/tests/main.nf.test:23]: Where would be more appropriate? This is simulated test-data provided by the tool-developers. I guess it is broadly "genomic"?

File: modules/nf-core/links/environment.yml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+---
+channels:
```
Comment (maxulysse) [inline modules/nf-core/links/environment.yml:2]:
```suggestion
# yaml-language-server: $schema=https://raw.githubusercontent.com/nf-core/modules/master/modules/environment-schema.json
channels:
```

File: modules/nf-core/links/meta.yml [hunk 4]
```diff
@@ -0,0 +1,175 @@
+---
```
Comment (maxulysse) [inline modules/nf-core/links/meta.yml:1]:
```suggestion
```

File: modules/nf-core/links/tests/main.nf.test [hunk 5]
```diff
@@ -0,0 +1,85 @@
+// nf-core modules test links
```
Comment (maxulysse) [inline modules/nf-core/links/tests/main.nf.test:1]:
```suggestion
```

File: modules/nf-core/links/tests/nextflow_links_test.config
Comment (maxulysse) [inline modules/nf-core/links/tests/nextflow_links_test.config:1]: why not call it just `nextflow.config`?

File: modules/nf-core/links/tests/main.nf.test [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                """
+            }
+        }
+
+        then {
+            assertAll(
+                { assert process.success },
+                { assert snapshot(
+                    file(process.out.log[0][1]).name,
+                    process.out.pairing_issues,
```
[the remark this answers is not in the document]
Comment (nschan) [reply modules/nf-core/links/tests/main.nf.test:38]: Probably with a test that actually has reads, and does not essentially scaffold against a single long sequence, but I am not 100% sure.

File: modules/nf-core/links/meta.yml [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+input:
+  - - meta:
+        type: map
+        description: |
+          Groovy Map containing sample information
+          e.g. `[ id:'sample1' ]`
+    - assembly:
+        type: file
+        description: (Multi-)fasta file containing the draft assembly
+        pattern: "*.{fa,fasta}"
```
[the remark this answers is not in the document]
Comment (nschan) [reply modules/nf-core/links/meta.yml]: I have come up with something that should handle zipped assembly, and / or reads.

File: modules/nf-core/links/tests/main.nf.test (hunk 2 above)
[the remark this answers is not in the document]
Comment (nschan) [reply modules/nf-core/links/tests/main.nf.test:21]: I moved this into `genomics/scaffolding`

File: modules/nf-core/links/tests/main.nf.test (hunk 6 above)
[the remark this answers is not in the document]
Comment (nschan) [reply modules/nf-core/links/tests/main.nf.test:36]: I have added more test-data, but none of those produce `pairing_distribution` . I am not sure if this is actually created, but it is listed as an output in the README. Maybe this was lost when they reimplemented perl LINKS in C++?

**Comment by nschan** [comment]
I have added some more tests using also the test data provided with longstitch. This has revealed that other files are also not stable when more realistic data is used, so for the tests with longstitch checksums for some additional files are not done, but file existence is asserted.
Anything that I missed from the review @maxulysse ?

**Review by maxulysse — APPROVED** [review]

File: modules/nf-core/links/tests/main.nf.test (hunk 6 above)
[the remark this answers is not in the document]
Comment (SPPearce) [reply modules/nf-core/links/tests/main.nf.test:36]: That's fine, you don't need to worry about all the optional files in my opinion.
